### PR TITLE
[Functions] Abstract repetitive code and add unit tests. #9502

### PR DESCRIPTION
**Pull request by golden-yang**
### Motivation
This PR is related to #9502 and [PR #9519](https://github.com/apache/pulsar/pull/9519).
In [PR #9519 ](https://github.com/apache/pulsar/pull/9519), another contributer has fixed the problem described in #9502, but there are still some areas that can be optimized. 

### Modifications
- Abstract a method and then reuse some code.
- Adding a unit test for the `restartFunctionInstances` method

And I am a newcomer to the pulsar community, do I need to recreate an issue?

### CODE REVIEW

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java [hunk 1]
```diff
@@ -452,25 +444,28 @@ public synchronized void restartFunctionInstances(String tenant, String namespac
...
-                            assignment.getInstance().getInstanceId());
-                    }
+                    restartFunctionByPulsarAdmin(assignment, tenant, namespace, functionName);
                 }
             }
         }
         return;
     }
 
+    private void restartFunctionByPulsarAdmin(Assignment assignment, String tenant, String namespace, String functionName)
```
Comment (eolivelli) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java:454]: what about
`restartFunctionUsingPulsarAdmin` ?

Not a big deal for me, just a suggestion

**Review by eolivelli — APPROVED** [review]
LGTM

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java (hunk 1 above)
Comment (golden-yang) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java:454]: I think your proposal is better. Thanks.

**Review by freeznet — APPROVED** [review]
LGTM

**Review by codelipenghui — APPROVED** [review]

**Comment by golden-yang** [comment]
Thanks for suggest from @xche. I found that in thread model and k8s model, the restart code is very similar in both cases, but not exactly the same.  I have fixed this problem and added a unit test for the k8s scene.Please check again @xche .

**Review by sijie — APPROVED** [review]

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java [hunk 2]
```diff
@@ -452,25 +444,43 @@ public synchronized void restartFunctionInstances(String tenant, String namespac
...
                 }
             }
         }
         return;
     }
 
+    /**
+     * Restart the entire function or restart a single instance of the function
+     */
+    private void restartFunctionUsingPulsarAdmin(Assignment assignment, String tenant, String namespace,
```
Comment (jerrypeng) [inline pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java:457]: I would recommend refactoring this method to the following: 

`
// restarts a particular instance of the function
private void internalRestartFunction(Assignment assignment, String tenant, String namespace, Integer instanceId)

// restarts the whole function
private void internalRestartFunction(Assignment assignment, String tenant, String namespace) {
  internalRestartFunction(assignment, tenant, namespace, null)
}
`

**Review by jerrypeng — CHANGES_REQUESTED** [review]
@golden-yang thanks for contributing.  I have added a comment.

**Review by jerrypeng — COMMENTED** [review]
Added a comment

File: pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java (hunk 2 above)
Comment (golden-yang) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java:457]: Thank you for your advice.
I realized that in my previous implementation, using the boolean value `restartEntireFunction ` to control function behavior was absolutely bad code.

How about split it into two methods, 'internalrestartinstance' and 'internalrestartfunction', will it be more intuitive？
// restarts a particular instance of the function
`private void internalRestartInstance(Assignment assignment, String tenant, String namespace, Integer instanceId){}`

// restarts the whole function
`private void internalRestartFunction(Assignment assignment, String tenant, String namespace) {}`

Do you think it's necessary? Or just take your advice directly.
@jerrypeng

Comment (sijie) [reply pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java:457]: @jerrypeng Do you want to check @golden-yang comment?

**Pull request merged**